### PR TITLE
Allow multiple parameters for nodes and add example with parametrized additive node

### LINKED ISSUE #98: Introduce multiplicative and additive parameters for each node
Currently the only way to introduce constants in the expressions is via the `ConstantFloat` or `Parameter` nodes. The values of the parameter nodes can be tuned by local search to specific values in order to increase the fitness of individuals.
However, there is a fundamental problem. As nodes can be used in more than one subsequent node as inputs, it can happen that a particular `Parameter` node feeds into two subsequent nodes. If one of them would require the parameter to be large, the other to be small, the gradient would point in opposite directions, thereby bringing the local search for this particular parameter to a halt, or at least slowing it down significantly. These can lead to difficult-to-escape local minima in the search landscape.

This can be avoided by additionally introducing parameters private to each node (e.g., Izzo et al., 2017). These private parameters are by definition not used by any other node and hence are never subject to "opposing" gradient information. For simplicity a first implementation would introduce two parameters, one that scales the output of each node and one that shifts it. By default these would have the values 1.0 and 0.0, respectively and hence not change the behavior of the node from the current implementation. I guess it could be helpful to introduce a mechanic that allows users to choose between updating only values of `Parameter` nodes, only private node parameters or both. As a side remark, this would allow the evolution of simple ANNs by introducing nodes that only apply an non-linearity to their input (e.g., CGPANNs; Khan et al., 2013).

> Izzo, D., Biscani, F., & Mereta, A. (2017). Differentiable Genetic Programming. In European Conference on Genetic Programming (pp. 35-51). Springer, Cham.
> Khan, M. M., Ahmad, A. M., Khan, G. M., & Miller, J. F. (2013). Fast Learning Neural Networks using Cartesian Genetic Programming. Neurocomputing, 121, 274-289.

### PR DESCRIPTION
**Pull request by jakobj**
With #178 this is now incredibly easy: one can just define new custom nodes that implement certain operations and may contain parameters that can be adapted via local search. This PR includes a fix for nodes that use more than one parameter and example for how to make use of the new power of #178 ;)

Should we include this parametrized add as a new default node? Or is is so simple now to define nodes that we don't need it?

closes #98

### CODE REVIEW

File: examples/example_parametrized_nodes.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
+import functools
```
Comment (mschmidt87) [inline examples/example_parametrized_nodes.py:1]: Please add more docstrings to get a nice documentation of this example in our sphinx-generated docs.

**Comment by jakobj** [comment]
good point! i've adapted the example to include proper documentation that can be properly rendered. please have another look.

File: examples/example_parametrized_nodes.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+
+# %%
+# We define a target function which contains numerical constants that
+# are not available as constants for the search and need to be found
+# by local search on parameterized nodes.
+
+
+def f_target(x):
+    return np.pi * x[:, 0] + 1.0 + np.pi
```
Comment (mschmidt87) [inline examples/example_parametrized_nodes.py:52]: why the `+ 1.0` here?

File: examples/example_parametrized_nodes.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+
+# %%
+# Then we define a differentiable(!) objective function for the
```
Comment (mschmidt87) [inline examples/example_parametrized_nodes.py:56]:
```suggestion
# Then we define a differentiable(!) inner objective function for the
```

File: examples/example_parametrized_nodes.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+# evolution. It accepts a torch tensor as input variable and uses
```
Comment (mschmidt87) [inline examples/example_parametrized_nodes.py:57]: Doesn't `inner_objective` take a function `f` as input?

Comment (mschmidt87) [inline examples/example_parametrized_nodes.py:57]:
```suggestion
# evolution. It accepts a function that returns a torch tensor as input variable and uses
```

File: examples/example_parametrized_nodes.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+obj = functools.partial(objective, seed=population_params["seed"])
+
+cgp.evolve(
+    pop, obj, ea, **evolve_params, print_progress=True, callback=recording_callback,
+)
+
+# %%
+# After the evolutionary search has ended, we print the expression
+# with the highest fitness and plot the progression of the search.
```
Comment (mschmidt87) [inline examples/example_parametrized_nodes.py:144]: Here, I would be interested to see the development of expressions over the course of the evolution. The fitness seems to increase very smoothly. Has the evolution found the right expression right in the beginning and then just keeps finetuning this over the course of the evolution?

File: examples/example_parametrized_nodes.py (hunk 2 above)
Comment (jakobj) [reply examples/example_parametrized_nodes.py:52]: no reason, just to avoid the same constant twice. i've changed this to `math.pi * x[:, 0] + math.e` now to avoid such confusion

File: examples/example_parametrized_nodes.py (hunk 5 above)
Comment (jakobj) [reply examples/example_parametrized_nodes.py:146]: i've improved the output of the example, this now shows that for the first few generations an expression without constants is used, then the correct form is found and the rest of the generations are busy finding the correct numerical values.

**Comment by jakobj** [comment]
thanks for your input @mschmidt87! i've included your suggestions, please have another look

**Comment by jakobj** [comment]
thanks! added the example to travis, sqashed, now waiting for green light from travis before merging
